src/apps/dictionary.man
.TH DICTIONARY 1 "AUGUST 2014" "Cainteoir Text-to-Speech" "Cainteoir Manuals"
.SH NAME
dictionary \- create, manage and test pronunciation dictionaries
.SH SYNOPSIS
.B dictionary [OPTION..]
.I DOCUMENT..
.SH DESCRIPTION
.B dictionary
is a command\-line program for working with pronunciation dictionaries.
.SH OPTIONS
.IP "-a ACCENT, --accent ACCENT"
Use the phoneme-to-phoneme rules for the specified accent to convert the phonemes
produced by --voice, --language or --source-dictionary. This is the name of a
supported accent (e.g. en-GB-rp or cmu), not a path to a phoneme-to-phoneme
conversion file.
.IP "-c, --comapre"
Compare the words in a pronunciation dictionary (--dictionary) with the
pronunciation generated by a text-to-speech engine (--voice or --language)
or found in another dictionary (--source-dictionary).
.IP "-d DICTIONARY, --dictionary=DICTIONARY"
Use the specified pronunciation dictionary file. Currently, only the Cainteoir
dictionary format is supported.
.IP "-D FORMAT, --as-dictionary FORMAT"
Print out the entries in the given dictionary format.
.IP "-f, --filter-words"
Create a new dictionary that contains words from the specified dictionary
that are in the document.
.IP "-h, --help"
Show a command-line option usage help message.
.IP "--ignore-syllable-breaks"
Ignore syllable breaks (/./) when comparing pronunciations.
.IP "-l LANGUAGE, --language=LANGUAGE"
Use a text-to-speech voice matching the specified language to pronounce words.
.IP "-L, --list"
Print out the entries in the dictionary. This is primarily used in tests.
.IP "-m, --mismatched"
Only display mis-pronounced words. This option implies the --compare option.
.IP "-M PHONEME_MAP, --phoneme-map PHONEME_MAP"
Use the phoneme-to-phoneme rules in PHONEME_MAP to convert the phonemes
produced by --voice, --language or --source-dictionary.
.IP "-n, --new-words"
Only add words that are not in the specified dictionary.
.IP "-p, --pronounce"
Pronounce the words from the specified document(s) using a text-to-speech
engine (--voice or --language) or another dictionary (--source-dictionary).
.IP "-P PHONEMESET, --phonemeset=PHONEMESET"
Use the specified phonemeset to transcribe the word pronunciations in.
.IP "-r, --ruleset"
Use the specified letter-to-phoneme rules to pronounce new words, or to compare
dictionary entries against.
.IP "-R, --resolve-say-as"
Dictionary entries marked as 'say-as' are resolved until a phoneme entry is
found. That is, if the entry 'ana' has the say-as value 'anna' then that entry
('anna') is looked up to use as the pronunciation for the entry ('ana'). This
implies the use of the --list option.
.IP "-s DICTIONARY, --source-dictionary=DICTIONARY"
Use the specified pronunciation dictionary file to use with --pronounce, --compare
or --mismatched as the source of pronunciations.
.IP "--syllable-stress"
Adjust the placement of stresses such that they are on syllable boundaries.
.IP "-t, --time"
Display how long it takes to run the dictionary command.
.IP "-v VOICE, --voice=VOICE"
Use the specified text-to-speech voice to pronounce words.
.IP "--vowel-stress"
Adjust the placement of stresses such that they are on vowels (e.g. eSpeak,
arpabet). Although eSpeak supports stresses placed on consonants, it places them
on vowels in its dictionaries and pronunciation rule files.
.SH SUPPORTED DICTIONARY FORMATS
.IP "cainteoir"
The Cainteoir Text-to-Speech dictionary format (see Cainteoir Dictionary Format
for a description of the format).
.IP "espeak"
The eSpeak dictionary format used by the _list, _listx and _extra files.
.SH CAINTEOIR DICTIONARY FORMAT
This is the format of the Cainteoir Text-to-Speech pronunciation dictionary.

A # character starts a comment. This comment extends to the end of the line.

Any blank lines are ignored.

In the entry format below, the spaces are actually TAB characters. One or more
TAB characters separate each section.
.SS Directives
.IP ".author AUTHOR"
Specifies the author of the pronunciation dictionary. Use one directive per
author.
.IP ".copyright COPYRIGHT"
A copyright statement.
.IP ".license LICENSE"
The license the pronunciation dictionary is assigned. For example, CC-BY-SA 3.0
or GPLv3.
.IP ".description DESCRIPTION"
A description of the pronunciation dictionary.
.IP ".phonemeset PHONEMESET"
The phonemeset used to transcribe pronunciations in the dictionary.
.IP ".import PATH"
Include the pronunciation entries from the dictionary located in PATH, relative
to the current dictionary file.
.SS Words
.IP "word /PRONUNCIATION/"
Specifies how the word is pronounced. The word must be in lower case, otherwise
it will not be recognised by Cainteoir Text-to-Speech.
.IP "word SAYAS"
Specifies that the pronunciation of the word is the SAYAS word. This allows
alternative spellings of a word to reference a cannonical spelling without
repeating the pronunciation for all spelling variants.
.SS Cardinal and Ordinal Numbers
These are special words that are used when describing how to pronounce cardinal
and ordinal numbers.
.IP "_Nx"
Used to denote words for the tens number column, where N is a number from 2 to 9.
For example, _3x is used to denote the word for thirty.
.IP "_10^2"
Used to denote the word for hundred.
.IP "_10^N"
Used to denote words for powers of 10 (where N is a multiple of 3) for the words
for thousand (_10^3), million (_10^6) and so on.
.IP "_andDD"
Used to denote a joining word (typically the word for "and") if a number requires
it. For example, the number 1024 is typically pronuonced "one thousand and twenty
four" in Brittish English.
.SH EXAMPLES
.IP "echo 'hello world' | dictionary -p -D cainteoir -l en"
Pronounce the words "hello" and "world" using an English eSpeak voice.
.IP "dictionary -pn -D cainteoir -l en -d words.dict WutheringHeights.epub"
Pronounce the words in "WutheringHeights.epub" that do not appear in "words.dict"
using an English eSpeak voice.
.IP "dictionary -tc -d espeak.dict -P ipa"
Compare the pronunciations of the words in the "espeak.dict" dictionary with the
default text-to-speech voice, transcribing the phonemes in IPA (International
Phonetic Alphabet). Print out the matching and mismatching entries, along with
the time taken to do the comparison.
.IP "dictionary -c -l en -d english.dict --syllable-stress --ignore-syllable-breaks"
Compare the pronunciation of the words in the "english.dict" dictionary with the
"en" (default) eSpeak voice. The pronunciations generated by eSpeak are modified to
place the stress on syllable boundaries and syllable breaks are not treated as
significant.
.IP "dictionary -v english -m -D espeak -d espeak.dict -P espeak/en"
Output words from espeak.dict that the "english" eSpeak voice mispronounces in
the format used by the en_extra eSpeak dictionary file.
.IP "echo 'hello world' | dictionary -f -d english.dict -D cainteoir"
Extract the pronunciation for 'hello' and 'world' from the "english.dict"
dictionary.
.SH AUTHOR
Reece H. Dunn <msclrhd@gmail.com>


NEWS

-------------------------------------------------------------------------------
VERSION 0.8 - Friday June 1st 2012 [Jun 2012]

  *  define an archive interface with a zip file implementation
  *  report the filename as a toc entry if no toc is present
  *  move the UTF-8 helpers to a common location
  *  optimize decoding of utf-8 and us-ascii encodings
  *  do not load a new decoder when switching to the same encoding
  *  switch document parsing from an event model to a reader model

applications:
  *  doc2doc: a simple command-line tool for converting documents from
     one format to another

document formats:
  *  pdf:     support PDF documents using the Poppler library
  *  (x)html: support extracting metadata from several meta tag names
  *  zip:     support reading generic zipped document collections
  *  ePub:    make the ePub parser use the archive interface

-------------------------------------------------------------------------------
VERSION 0.7 - Friday March 30th 2012 [Mar 2012]

  *  use BCP 47 for interpreting and comparing language tags
  *  translate the language and region names using the iso-codes package
  *  support for espeak using installed mbrola voices

document formats:
  *  (x)html: report h1..h6 sections as table of content items
  *  email:   improve email content detection and support newsgroup information
  *  rdf/xml: support property attributes on an empty property element
  *  xml:     support entities in DTD blocks

bug fixes:
  *  fix detection of html that is valid xml, but is not xhtml
  *  improve detection of control characters when identifying binary data

cleanup/refactoring:
  *  simplified and cleaned up the RDF model and query APIs
  *  use std::shared_ptr consistently over std::tr1::shared_ptr
  *  use std::make_shared instead of allocating the object explicitly
  *  remove the libxml2 dependency and make all xml-based parsers use xmlreader

-------------------------------------------------------------------------------
VERSION 0.6 -- Sunday January 29th 2012 [Jan 2012]

  *  languages.rdf -- represent ISO-639-2/B, ISO-639-2/T, ISO-639-3, ISO-639-5,
     espeak and wikimedia language codes;
  *  add document format implementation status documentation.

xmlreader xml parsing support:
  *  parse entities in attribues;
  *  support single-quote and html-style attributes;
  *  parse xml namespaces;
  *  unify xml and rdf namespace/uri support;
  *  support element/attribute mapping to context entry objects.

document parsers:
  *  epub  -- support epub 3 @refines and @datatype metadata;
  *  html  -- support all known implicit close tags and recognise more elements;
  *  ocf   -- use the xmlreader api and expose a reader api;
  *  opf   -- use the xmlreader api.

bug fixes:
  *  release the file handle when finished recording ogg files.

-------------------------------------------------------------------------------
VERSION 0.5 -- Monday November 21st 2011 [Nov 2011]

  *  use the shared-mime-info database for MIME type detection.
  *  use iconv for character encoding conversions.
  *  support all Content-Transfer-Encoding types for MIME headers, including base64.
  *  improve audio error handling during reading/recording.
  *  support more language code mappings for new espeak voices.
  *  support UND as a language identifier (for Calibri eBooks).
  *  support XML encoded HTML without an associated xmlns.

-------------------------------------------------------------------------------
VERSION 0.4 -- Tuesday July 26th 2011 [Jul 2011]

file formats:
  *  support ncx-based table of content in epub documents;
  *  support epub 3.0 metadata;
  *  basic support for SSML.

cainteoir application and engine:
  *  support listing a document's table of contents and read/record a
     specified content range;
  *  support setting the voice reading speed, pitch and volume.

-------------------------------------------------------------------------------
VERSION 0.3 -- Friday July 1st 2011 [Jun 2011]

file formats:
  *  support for single-file HTML pages (MHTML, MHT);
  *  support for RTF documents;
  *  enhanced HTML support.

bug fixes:
  *  don't segfault if the OPF document is missing from an epub file.

cainteoir:
  *  improved total reading time estimating heuristics.

-------------------------------------------------------------------------------
VERSION 0.2 -- Sunday May 29th 2011 [May 2011]

file formats:
  *  support email (RFC822) mbox format;
  *  basic support for RTF documents.

metadata:
  *  add rdfs:domain, rdfs:range and OWL markup to the text-to-speech schema;
  *  install the text-to-speech RDF schema used by the Cainteoir Engine;
  *  extract language and title metadata from HTML documents;
  *  extract metadata from email From and Subject headers.

cainteoir:
  *  list supported document and audio formats in the --help output;
  *  don't output progress when writing audio data to stdout;
  *  don't crash when using the default voice (no dc:language metadata).

-------------------------------------------------------------------------------
VERSION 0.1 -- Wednesday April 27th 2011 [Apr 2011]

cainteoir:
  *  command-line text-to-speech program;
  *  support for using espeak text-to-speech voices;
  *  listen to text, html and epub documents on the command line;
  *  record text, html and epub documents to ogg/vorbis and wave audio;
  *  see the reading/recording progress and (estimated) total time;
  *  preserve metadata from document to audio (currently epub to ogg/vorbis).

metadata:
  *  extract metadata from epub, smil, rdf/xml and other documents;
  *  generate RDF N-Triples
  *  generate RDF Turtle statements
  *  generate VorbisComment metadata


src/apps/dictionary.cpp
/* Test for the exception dictionary API.
 *
 * Copyright (C) 2013-2014 Reece H. Dunn
 *
 * This file is part of cainteoir-engine.
 *
 * cainteoir-engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cainteoir-engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cainteoir-engine.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include "i18n.h"
#include "options.hpp"

#include <cainteoir/document.hpp>
#include <cainteoir/engines.hpp>
#include <cainteoir/text.hpp>
#include <cainteoir/language.hpp>
#include <cainteoir/path.hpp>
#include <cainteoir/stopwatch.hpp>
#include <stdexcept>

namespace rdf = cainteoir::rdf;
namespace tts = cainteoir::tts;
namespace ipa = cainteoir::ipa;

enum class mode_type
{
	from_document,
	list_entries,
	resolve_say_as_entries,
	pronounce_entries,
	compare_entries,
	mismatched_entries,
};

enum class word_mode_type
{
	merge,
	only_in_document, // new words
	in_dictionary_and_document,
};

struct dictionary_phoneme_reader : public tts::phoneme_reader
{
	dictionary_phoneme_reader(const char *aDictionary);

	void reset(const std::shared_ptr<cainteoir::buffer> &aBuffer);

	bool read();
private:
	tts::dictionary mDictionary;
	ipa::phonemes mPhonemes;
	ipa::phonemes::const_iterator mCurrent;
	ipa::phonemes::const_iterator mLast;
};

dictionary_phoneme_reader::dictionary_phoneme_reader(const char *aDictionary)
{
	auto reader = tts::createCainteoirDictionaryReader(aDictionary);
	while (reader->read())
		mDictionary.add_entry(reader->word, reader->entry);
}

void dictionary_phoneme_reader::reset(const std::shared_ptr<cainteoir::buffer> &aBuffer)
{
	mPhonemes.clear();
	if (!mDictionary.pronounce(aBuffer, {}, mPhonemes))
		throw tts::phoneme_error("the word is not in the dictionary");

	mCurrent = mPhonemes.begin();
	mLast = mPhonemes.end();
}

bool dictionary_phoneme_reader::read()
{
	if (mCurrent == mLast)
		return false;
	*(ipa::phoneme *)this = *mCurrent++;
	return true;
}

static bool matches(const ipa::phonemes &a, const ipa::phonemes &b, bool ignore_syllable_breaks)
{
	auto first1 = a.begin(), last1 = a.end();
	auto first2 = b.begin(), last2 = b.end();
	while (first1 != last1 && first2 != last2)
	{
		if (ignore_syllable_breaks)
		{
			if (*first1 == ipa::syllable_break) { ++first1; continue; }
			if (*first2 == ipa::syllable_break) { ++first2; continue; }
		}

		if (*first1 != *first2) return false;
		++first1;
		++first2;
	}
	return first1 == last1 && first2 == last2;
}

static bool is_variant(const cainteoir::buffer &s)
{
	for (auto c : s)
	{
		if (c == '@') return true;
	}
	return false;
}

static bool pronounce(tts::dictionary &dict,
                      const std::shared_ptr<cainteoir::buffer> &word,
                      std::shared_ptr<tts::phoneme_reader> &rules,
                      std::shared_ptr<tts::dictionary_formatter> &formatter,
                      std::shared_ptr<tts::phoneme_writer> &writer,
                      const char *phonemeset,
                      tts::stress_type stress,
                      bool ignore_syllable_breaks,
                      mode_type mode)
{
	ipa::phonemes phonemes;
	if (mode != mode_type::pronounce_entries)
	{
		if (!dict.pronounce(word, {}, phonemes))
		{
			auto &entry = dict.lookup(word);
			switch (entry.type)
			{
			case tts::dictionary::say_as:
				fprintf(stderr, "error: cannot resolve '%s' from '%s'.\n",
				        entry.text->str().c_str(),
				        word->str().c_str());
				break;
			case tts::dictionary::no_match:
				fprintf(stderr, "error: no entry for '%s'.\n", word->str().c_str());
				break;
			}
			return false;
		}
	}

	ipa::phonemes pronounced;
	try
	{
		rules->reset(word);
		while (rules->read())
			pronounced.push_back(*rules);
	}
	catch (const tts::phoneme_error &e)
	{
		fprintf(stderr, "cannot pronounce '%s': %s\n", word->str().c_str(), e.what());
		return false;
	}

	tts::make_stressed(pronounced, stress);

	bool match = matches(pronounced, phonemes, ignore_syllable_breaks);
	if (mode == mode_type::mismatched_entries && match)
		return true;

	if (mode == mode_type::compare_entries)
	{
		formatter->write_phoneme_entry(word, writer, phonemes, " ... ");
		if (match)
		{
			fprintf(stdout, "matched\n");
		}
		else
		{
			fprintf(stdout, "mismatched; got /");
			for (auto p : pronounced)
				writer->write(p);
			writer->flush();
			fprintf(stdout, "/\n");
		}
	}
	else if (mode == mode_type::mismatched_entries)
		formatter->write_phoneme_entry(word, writer, phonemes);
	else
		formatter->write_phoneme_entry(word, writer, pronounced);

	return match;
}

static void pronounce(tts::dictionary &dict,
                      std::shared_ptr<tts::phoneme_reader> rules,
                      std::shared_ptr<tts::dictionary_formatter> &formatter,
                      std::shared_ptr<tts::phoneme_writer> writer,
                      const char *phonemeset,
                      tts::stress_type stress,
                      bool ignore_syllable_breaks,
                      mode_type mode)
{
	writer->reset(stdout);

	int matched = 0;
	int entries = 0;

	for (auto &entry : dict)
	{
		if (is_variant(*entry.first)) continue;

		if (pronounce(dict, entry.first,
		              rules, formatter, writer, phonemeset, stress, ignore_syllable_breaks, mode))
			++matched;
		++entries;
	}

	fflush(stdout);

	if (mode == mode_type::compare_entries)
	{
		fprintf(stderr, "... matched: %d (%.0f%%)\n", matched, (float(matched) / entries * 100.0f));
		fprintf(stderr, "... entries: %d\n", entries);
	}
}

static uint32_t from_document(tts::dictionary &base_dict,
                              tts::dictionary &dict,
                              const char *filename,
                              bool silent,
                              word_mode_type word_mode)
{
	if (!silent)
		fprintf(stdout, "reading %s\n", (filename == nullptr) ? "<stdin>" : filename);

	rdf::graph metadata;
	auto reader = cainteoir::createDocumentReader(filename, metadata, std::string());
	if (!reader)
	{
		fprintf(stderr, "unsupported document format for file \"%s\"\n",
		        (filename == nullptr) ? "<stdin>" : filename);
		return 0;
	}

	std::shared_ptr<tts::phoneme_reader> phonemeset;
	uint32_t words = 0;
	auto text = tts::create_text_reader(reader);
	while (text->read()) switch (text->event().type)
	{
	case tts::word_uppercase:
	case tts::word_lowercase:
	case tts::word_capitalized:
	case tts::word_mixedcase:
	case tts::word_script:
		{
			auto &match = base_dict.lookup(text->event().text);
			switch (match.type)
			{
			case tts::dictionary::no_match:
				switch (word_mode)
				{
				case word_mode_type::merge:
				case word_mode_type::only_in_document:
					dict.add_entry(text->event().text, { text->event().text });
					++words;
					break;
				default:
					fprintf(stderr, "error: cannot find '%s' in the dictionary\n", text->event().text->str().c_str());
					break;
				}
				break;
			case tts::dictionary::say_as:
				if (word_mode == word_mode_type::in_dictionary_and_document)
				{
					ipa::phonemes pronunciation;
					if (base_dict.pronounce(text->event().text, {}, pronunciation))
					{
						dict.add_entry(text->event().text, { pronunciation });
						++words;
					}
				}
				break;
			case tts::dictionary::phonemes:
				if (word_mode == word_mode_type::in_dictionary_and_document)
				{
					dict.add_entry(text->event().text, match );
					++words;
				}
				break;
			}
		}
		break;
	}

	return words;
}

int main(int argc, char ** argv)
{
	try
	{
		tts::stress_type stress = tts::stress_type::as_transcribed;
		mode_type mode = mode_type::from_document;
		bool time = false;
		word_mode_type word_mode = word_mode_type::merge;
		bool ignore_syllable_breaks = false;
		const char *voicename = nullptr;
		const char *language = nullptr;
		const char *ruleset = nullptr;
		const char *dictionary = nullptr;
		const char *dictionary_format = nullptr;
		const char *phonemeset = "ipa";
		const char *source_dictionary = nullptr;
		const char *phoneme_map = nullptr;
		const char *accent = nullptr;

		const option_group general_options = { nullptr, {
			{ 't', "time", bind_value(time, true),
			  i18n("Time how long it takes to complete the action") },
			{ 'd', "dictionary", dictionary, "DICTIONARY",
			  i18n("Use the words in DICTIONARY") },
			{ 'M', "phoneme-map", phoneme_map, "PHONEME_MAP",
			  i18n("Use PHONEME_MAP to convert phonemes (e.g. accent conversion)") },
			{ 'a', "accent", accent, "ACCENT",
			  i18n("Use ACCENT to convert phonemes to the specified accent") },
		}};

		const option_group output_options = { i18n("Output:"), {
			{ 'D', "as-dictionary", dictionary_format, "FORMAT",
			  i18n("List the entries in the given dictionary format") },
			{ 'P', "phonemeset", phonemeset, "PHONEMESET",
			  i18n("Use PHONEMESET to transcribe phoneme entries (default: ipa)") },
			{ 'f', "filter-words", bind_value(word_mode, word_mode_type::in_dictionary_and_document),
			  i18n("Only use dictionary entries that are in the input documents") },
			{ 'n', "new-words", bind_value(word_mode, word_mode_type::only_in_document),
			  i18n("Only use words not in the loaded dictionary") },
		}};

		const option_group stress_options = { i18n("Stress:"), {
			{ 0, "vowel-stress", bind_value(stress, tts::stress_type::vowel),
			  i18n("Place the stress on vowels (e.g. espeak, arpabet)") },
			{ 0, "syllable-stress", bind_value(stress, tts::stress_type::syllable),
			  i18n("Place the stress on syllable boundaries") },
			{ 0, "ignore-syllable-breaks", bind_value(ignore_syllable_breaks, true),
			  i18n("Ignore syllable breaks (/./) when comparing pronunciations.") },
		}};

		const option_group action_options = { i18n("Action:"), {
			{ 'L', "list", bind_value(mode, mode_type::list_entries),
			  i18n("List the entries in the dictionary") },
			{ 'R', "resolve-say-as", bind_value(mode, mode_type::resolve_say_as_entries),
			  i18n("Look up say-as entries in the dictionary (implies --list)") },
			{ 'p', "pronounce", bind_value(mode, mode_type::pronounce_entries),
			  i18n("Use the Pronunciation options to pronounce the entries") },
			{ 'c', "compare", bind_value(mode, mode_type::compare_entries),
			  i18n("Compare dictionary and Pronunciation options pronunciation") },
			{ 'm', "mismatched", bind_value(mode, mode_type::mismatched_entries),
			  i18n("Only display mismatched pronunciations (implies --compare)") },
		}};

		const option_group pronunciation_options = { i18n("Pronunciation:"), {
			{ 's', "source-dictionary", source_dictionary, "DICTIONARY",
			  i18n("Use the pronunciation of the words in DICTIONARY") },
			{ 'r', "ruleset", ruleset, "RULESET",
			  i18n("Use the RULESET pronunciation rule file") },
			{ 'v', "voice", voicename, "VOICE",
			  i18n("Use the TTS voice named VOICE") },
			{ 'l', "language", language, "LANG",
			  i18n("Use a TTS voice that speaks the language LANG") },
		}};

		const std::initializer_list<const char *> usage = {
			i18n("dictionary [OPTION..] DOCUMENT.."),
		};

		if (!parse_command_line({ general_options, output_options, stress_options, action_options, pronunciation_options }, usage, argc, argv))
			return 0;

		auto writer = tts::createPhonemeWriter(phonemeset);
		cainteoir::stopwatch timer;

		tts::dictionary base_dict;
		tts::dictionary dict;

		uint32_t words = 0;
		if (dictionary != nullptr)
		{
			auto reader = tts::createCainteoirDictionaryReader(dictionary);
			while (reader->read())
				base_dict.add_entry(reader->word, reader->entry);
			if (word_mode == word_mode_type::merge)
			{
				dict  = base_dict;
				words = dict.size();
			}
		}

		if (dictionary == nullptr || word_mode != word_mode_type::merge)
		{
			if (argc == 0)
				words += from_document(base_dict, dict, nullptr, dictionary_format != nullptr, word_mode);
			else for (int i = 0; i != argc; ++i)
				words += from_document(base_dict, dict, argv[i], dictionary_format != nullptr, word_mode);
		}

		std::shared_ptr<tts::dictionary_formatter> formatter;
		if (!dictionary_format)
			formatter = tts::createDictionaryEntryFormatter(stdout);
		else if (!strcmp(dictionary_format, "cainteoir"))
			formatter = tts::createCainteoirDictionaryFormatter(stdout);
		else if (!strcmp(dictionary_format, "espeak"))
			formatter = tts::createEspeakDictionaryFormatter(stdout);
		else
		{
			fprintf(stderr, "unsupported dictionary format \"%s\"\n", dictionary_format);
			return 0;
		}

		if (mode == mode_type::from_document && dictionary_format != nullptr)
			mode = mode_type::list_entries;

		switch (mode)
		{
		case mode_type::list_entries:
		case mode_type::resolve_say_as_entries:
			writer->reset(stdout);
			tts::formatDictionary(dict, formatter, writer, mode == mode_type::resolve_say_as_entries);
			break;
		case mode_type::pronounce_entries:
		case mode_type::compare_entries:
		case mode_type::mismatched_entries:
			if (source_dictionary != nullptr)
			{
				std::shared_ptr<tts::phoneme_reader> rules = std::make_shared<dictionary_phoneme_reader>(source_dictionary);
				if (phoneme_map)
					rules = tts::createPhonemeToPhonemeConverter(phoneme_map, rules);
				if (accent)
					rules = tts::createAccentConverter(accent, rules);
				pronounce(dict, rules, formatter, writer, phonemeset, stress, ignore_syllable_breaks, mode);
			}
			else if (ruleset != nullptr)
			{
				auto rules = tts::createPronunciationRules(ruleset);
				if (!rules.get())
				{
					fprintf(stderr, "cannot load letter-to-phoneme rule file \"%s\"\n", argv[1]);
					return 0;
				}
				if (phoneme_map)
					rules = tts::createPhonemeToPhonemeConverter(phoneme_map, rules);
				if (accent)
					rules = tts::createAccentConverter(accent, rules);
				pronounce(dict, rules, formatter, writer, phonemeset, stress, ignore_syllable_breaks, mode);
			}
			else
			{
				rdf::graph metadata;
				tts::engines engine(metadata);
				if (voicename)
				{
					const rdf::uri *ref = tts::get_voice_uri(metadata, rdf::tts("name"), voicename);
					if (ref)
						engine.select_voice(metadata, *ref);
				}
				else if (language)
				{
					const rdf::uri *ref = tts::get_voice_uri(metadata, rdf::dc("language"), language);
					if (ref)
						engine.select_voice(metadata, *ref);
				}
				auto rules = engine.pronunciation();
				if (phoneme_map)
					rules = tts::createPhonemeToPhonemeConverter(phoneme_map, rules);
				if (accent)
					rules = tts::createAccentConverter(accent, rules);
				pronounce(dict, rules, formatter, writer, phonemeset, stress, ignore_syllable_breaks, mode);
			}
			break;
		case mode_type::from_document:
			fflush(stdout);
			fprintf(stderr, "... words:   %d\n", words);
			fprintf(stderr, "... indexed: %zd\n", dict.size());
			break;
		}

		if (time)
			fprintf(stderr, "... time:    %G\n", timer.elapsed());
	}
	catch (std::runtime_error &e)
	{
		fprintf(stderr, "error: %s\n", e.what());
	}

	return 0;
}
